### PR TITLE
OTT-864: As a prebid server, add prometheus metrics for invalid fetched floors JSON

### PR DESCRIPTION
**Pull request by PubMatic-OpenWrap**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -28,7 +28,7 @@ require (
 	github.com/prometheus/client_model v0.2.0
 	github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
 	github.com/rs/cors v1.8.2
-	github.com/sergi/go-diff v1.2.0 // indirect
+	github.com/sergi/go-diff v1.3.1 // indirect
```
Comment (pm-jaydeep-mohite) [inline go.mod:31]: do we need it ?

File: metrics/prometheus/prometheus_test.go [hunk 2]
```diff
@@ -1979,3 +1979,42 @@ func TestRecordAdsCertSignTime(t *testing.T) {
...
+	}
+	testCases := []struct {
+		description string
+		in          testIn
+		out         testOut
+	}{
+		{
+			description: "record dynamic fetch failure",
+			in: testIn{
+				pubid:  "1010",
```
Comment (pm-jaydeep-mohite) [inline metrics/prometheus/prometheus_test.go:1998]: is code correct ?

File: go.mod [hunk 3]
```diff
@@ -28,7 +28,7 @@ require (
 	github.com/prometheus/client_model v0.2.0
```
Comment (Pubmatic-Dhruv-Sonone) [inline go.mod:28]: Run go mod tidy

File: floors/fetcher.go [hunk 4]
```diff
@@ -193,10 +196,11 @@ func (f *PriceFloorFetcher) Fetcher() {
 	}
 }
 
-func fetchAndValidate(configs config.AccountFloorFetch) (*openrtb_ext.PriceFloorRules, int) {
+func fetchAndValidate(configs config.AccountFloorFetch, metricEngine metrics.MetricsEngine) (*openrtb_ext.PriceFloorRules, int) {
 
 	floorResp, maxAge, err := fetchFloorRulesFromURL(configs)
 	if err != nil {
+		metricEngine.RecordDynamicFetchFailure(configs.AccountID, "1")
```
Comment (Pubmatic-Dhruv-Sonone) [inline floors/fetcher.go:203]: What does  "1" signify? Can we make this a constant

File: floors/fetcher.go [hunk 5]
```diff
@@ -134,7 +137,7 @@ func (f *PriceFloorFetcher) Fetch(configs config.AccountPriceFloors) (*openrtb_e
 
 func (f *PriceFloorFetcher) worker(configs config.AccountFloorFetch) {
 
-	floorData, fetchedMaxAge := fetchAndValidate(configs)
+	floorData, fetchedMaxAge := fetchAndValidate(configs, f.metricEngine)
```
Comment (pm-nikhil-vaidya) [inline floors/fetcher.go:140]: here new parameter to the function was added, need to modify UT for it as well
Add UT for the changes in fetcher code

File: floors/floors.go [hunk 6]
```diff
@@ -137,8 +137,9 @@ func resolveFloors(account config.Account, bidRequestWrapper *openrtb_ext.Reques
 	var floorsJson *openrtb_ext.PriceFloorRules
 
 	reqFloor := extractFloorsFromRequest(bidRequestWrapper)
-	if reqFloor != nil && reqFloor.Location != nil && reqFloor.Location.URL != "" {
+	if reqFloor != nil && reqFloor.Location != nil && len(reqFloor.Location.URL) > 0 {
 		account.PriceFloors.Fetch.URL = reqFloor.Location.URL
+		account.PriceFloors.Fetch.AccountID = account.ID
```
Comment (pm-nikhil-vaidya) [inline floors/floors.go:142]: From which part of the request we are reading the account ID?

File: floors/fetcher_test.go [hunk 7]
```diff
@@ -804,7 +805,7 @@ func TestFetcherDataPresentInCache(t *testing.T) {
 
 func TestFetcherDataNotPresentInCache(t *testing.T) {
 
-	fectherInstance := NewPriceFloorFetcher(2, 5, 5, 20)
+	fectherInstance := NewPriceFloorFetcher(2, 5, 5, 20, &metricsConf.NilMetricsEngine{})
```
Comment (pm-nikhil-vaidya) [inline floors/fetcher_test.go:808]: Create a test case where metrics are actually recorded

Comment (pm-nikhil-vaidya) [reply floors/fetcher_test.go:808]: Instead of passing nil metric engine

**Review by pm-jaydeep-mohite — APPROVED** [review]
LGTM

**Pull request merged**